### PR TITLE
Copy pypi release automation from `instructlab/instructlab` repo

### PR DESCRIPTION
**Pull request by russellb**
commit 1f01ce86d59917d6800e1b9d5d4a84331eb5e25a
Author: Russell Bryant <rbryant@redhat.com>
Date:   Wed Jun 12 10:22:56 2024 -0400

    Copy pypi release automation from `instructlab/instructlab` repo
    
    This also includes hardening changes from https://github.com/instructlab/schema/blob/52d22f4a0ebdc9b057bbd7c4e525df0c84d5d5cd/.github/workflows/pypi.yml
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

**Comment by nathan-weinberg** [comment]
Adding myself so I can do the same for https://github.com/instructlab/eval

**Comment by bjhargrave** [comment]
You may want to look at https://github.com/instructlab/schema/pull/27/files#diff-1baea138c19b06c7531655c5e977dbbcca5bd9147709625e2ee40984905502c5 which is a hardened version from the cli repo.

**Comment by russellb** [comment]
> > You may want to look at https://github.com/instructlab/schema/pull/27/files#diff-1baea138c19b06c7531655c5e977dbbcca5bd9147709625e2ee40984905502c5 which is a hardened version from the cli repo.
> 
> Thanks - I copied what was currently in the CLI repo. I'll look at what you have

done!

File: .github/workflows/pypi.yaml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    # - upload artifacts to GHA
+    build-package:
+        name: Build and check packages
+        runs-on: ubuntu-latest
+        steps:
+            - name: "Harden Runner"
+              uses: step-security/harden-runner@17d0e2bd7d51742c71671bd19fa12bdc9d40a3d6 # v2.8.1
+              with:
+                  egress-policy: audit # TODO: change to 'egress-policy: block' after couple of runs
+
```
Comment (nathan-weinberg) [inline .github/workflows/pypi.yaml:43]: nit: extra newline here

**Review by nathan-weinberg — COMMENTED** [review]
@russellb curious as I will be doing the same for eval - did you need to reserve anything/do any other outside steps for PyPI?

**Comment by tiran** [comment]
You need to follow these instructions: https://docs.pypi.org/trusted-publishers/creating-a-project-through-oidc/

File: .github/workflows/pypi.yaml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            - name: "Build and Inspect"
+              uses: hynek/build-and-inspect-python-package@b4fc3f6ba2b3da04f09659be99e2a29fb6146a61 # v2.6.0
+
+    # push to Test PyPI on
+    # - a new GitHub release is published
+    # - a PR is merged into main branch
+    publish-test-pypi:
+        name: Publish packages to test.pypi.org
+        # environment: publish-test-pypi
+        if: ${{ false && (github.repository_owner == 'instructlab') && ((github.event.action == 'published') || ((github.event_name == 'push') && (github.ref == 'refs/heads/main'))) }}
```
Comment (bjhargrave) [inline .github/workflows/pypi.yaml:60]: Reminder, you will need to remove `false &&` to activate this job when ready.

File: .github/workflows/pypi.yaml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+              uses: pypa/gh-action-pypi-publish@81e9d935c883d0b210363ab89cf05f3894778450 # v1.8.14
+              with:
+                  repository-url: https://test.pypi.org/legacy/
+
+    # push to Production PyPI on
+    # - a new GitHub release is published
+    publish-pypi:
+        name: Publish release to pypi.org
+        # environment: publish-pypi
+        if: ${{ false && (github.repository_owner == 'instructlab') && (github.event.action == 'published') }}
```
Comment (bjhargrave) [inline .github/workflows/pypi.yaml:90]: Reminder, you will need to remove `false &&` to activate this job when ready.

**Comment by russellb** [comment]
> You need to follow these instructions: https://docs.pypi.org/trusted-publishers/creating-a-project-through-oidc/

perfect, thanks!  This part is done.

**Review by nathan-weinberg — APPROVED** [review]

File: .github/workflows/pypi.yaml (hunk 2 above)
Comment (russellb) [reply .github/workflows/pypi.yaml]: done!

File: .github/workflows/pypi.yaml (hunk 3 above)
Comment (russellb) [reply .github/workflows/pypi.yaml]: done!

**Pull request merged**